api/kv/value/value.go

```diff
@@ -130,6 +135,52 @@ func Uint(v uint) Value {
 	return Uint64(uint64(v))
 }
 
+func isAllowedType(object interface{}) bool {
+	switch reflect.TypeOf(object).Kind() {
+	case reflect.Array, reflect.Slice,
+		reflect.Float32, reflect.Float64,
+		reflect.Int, reflect.Int32, reflect.Int64,
+		reflect.Uint, reflect.Uint32, reflect.Uint64,
+		reflect.String:
+		return true
+	}
+	return false
+}
+
+// isArrayHomogenous checks if array elements are same type
+func isArrayHomogenous(collection interface{}) bool {
```

Comment:
(I'll say that I have misgivings about the specification in this area. I don't see why the spec needs to state that values are homogenous. I think exporters should deal with mixed values by coercing to a slice of strings, they can deal with nested and nil values that way too!)

I'd be inclined to add a special case that avoids the for loop here when the value type is homogenous. If the collection is a Slice or an Array and the value type is not an allowed type, then the collection is invalid. I think we only need to test every item when you have a slice or array of `interface{}`.

Reply:
Agreed, but based on the conversation with @MrAlias we should support multi dimensional arrays too, that's I took approach of making method generic and evaluate type of each item during run time. Otherwise we can add method which take homogeneous type e.g `func Integars([]int{})`

Reply:
Hmm, I might have failed to make myself clear. When I was saying "[I think you could try to leverage the Go type system a bit more which already guarantees slice/array homogeneity](https://github.com/open-telemetry/opentelemetry-go/issues/718#issuecomment-637628515)" I was thinking more along the lines of what @jmacd is outlining. I think he is correct in that this validation should probably be reserved for just `[]interface{}` types as the type system allows up to make an assumption on homogeneity and we could exit early (or skip this function call?).

Reply:
I was not aware of the specification supporting anything other than arrays of scalar values. The cost of this test is not worth the value is provides, IMO, because ultimately I do not believe we should be in the business of telling users what is and is not valid data when they're trying to observe a system. 

(If multi-dimensional arrays are permitted, then I think we should just remove all validity tests. Remember that this testing will be done even when a span is sampled out. It's much better to move any kind of data format checking into an exporter and let the exporter to Deal With It.)

Reply:
> The cost of this test is not worth the value is provides, IMO, because ultimately I do not believe we should be in the business of telling users what is and is not valid data when they're trying to observe a system.

:100: 

> (If multi-dimensional arrays are permitted, then I think we should just remove all validity tests. Remember that this testing will be done even when a span is sampled out. It's much better to move any kind of data format checking into an exporter and let the exporter to Deal With It.)

I don't think they are intended to be permitted (I was assuming a loose interpretation of the specification where they don't explicitly define "primitive types"). Regardless, I think you are correct: assume the instrumenter send valid values, support them, and have the exporters deal with it.